### PR TITLE
Added dockerfile template

### PR DESCRIPTION
**Pull request by FlorianWilhelm**
Add a Dockerfile that gives an example of how to easily dockerize for instance a model that is accessible via a console_script.

### CODE REVIEW

**Comment by FlorianWilhelm** [comment]
@abravalheri, what do you think about that approach?

File: src/pyscaffoldext/dsproject/templates/Dockerfile.template [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+# build package
+RUN tox -e build
+RUN mamba env export -n ${name} -f environment.docker.yml
+
+FROM condaforge/miniforge3 AS runner
+WORKDIR /app
+
+# Copy caches, environment.docker.yml and built package
+COPY --from=builder /root/.cache/pip /root/.cache/pip
```
Comment (abravalheri) [inline src/pyscaffoldext/dsproject/templates/Dockerfile.template:30]: If we switch to the user `app`, this cache (`root/.cache/pip`) will lose it effectiveness, right?

**Comment by abravalheri** [comment]
Thank you very much @FlorianWilhelm, this approach seems very elegant!
I added a few remarks to the code (some of them are doubts I have myself), but please feel free to proceed with the code you already have.

File: src/pyscaffoldext/dsproject/templates/Dockerfile.template [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#  - save environment for runner
+# * runner steps:
+#  - recreate the same environment and install built package
+#  - optionally execute provided console_script in ENTRYPOINT
+#
+# Alternatively, remove builder steps, take `environment.docker.yml` from your repo
+# and `pip install ${name}` using an artifact store. Faster and more robust.
+
+FROM condaforge/miniforge3 AS builder
+WORKDIR /tmp
```
[the remark this answers is not in the document]
Comment (FlorianWilhelm) [reply src/pyscaffoldext/dsproject/templates/Dockerfile.template]: Changed it to `/root`

File: src/pyscaffoldext/dsproject/templates/Dockerfile.template [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#  - create environment and build package
+#  - save environment for runner
+# * runner steps:
+#  - recreate the same environment and install built package
+#  - optionally execute provided console_script in ENTRYPOINT
+#
+# Alternatively, remove builder steps, take `environment.docker.yml` from your repo
+# and `pip install ${name}` using an artifact store. Faster and more robust.
+
+FROM condaforge/miniforge3 AS builder
```
[the remark this answers is not in the document]
Comment (FlorianWilhelm) [reply src/pyscaffoldext/dsproject/templates/Dockerfile.template]: Good point.

File: src/pyscaffoldext/dsproject/templates/Dockerfile.template [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+RUN conda install mamba -n base
+RUN mamba env create -f environment.yml
+# RUN pip install -e . # not needed since it's in environment.yml
+SHELL ["mamba", "run", "-n", "${name}", "/bin/bash", "-c"]
+
+# build package
+RUN tox -e build
+RUN mamba env export -n ${name} -f environment.docker.yml
+
+FROM condaforge/miniforge3 AS runner
```
[the remark this answers is not in the document]
Comment (FlorianWilhelm) [reply src/pyscaffoldext/dsproject/templates/Dockerfile.template]: good point.

**Comment by abravalheri** [comment]
By the CI logs (`LookupError: setuptools-scm was unable to detect version for '/root'.
`), it seems that the `.git` directory is missing...

A quick and dirty workaround is to define `SETUPTOOLS_SCM_PRETEND_VERSION` as an env var...

**Comment by FlorianWilhelm** [comment]
> By the CI logs (`LookupError: setuptools-scm was unable to detect version for '/root'. `), it seems that the `.git` directory is missing...
> 
> A quick and dirty workaround is to define `SETUPTOOLS_SCM_PRETEND_VERSION` as an env var...
> 
> But then, the generated containers will not benefit from `setuptools-scm`...
> 
> I wonder if in the `build` phase would it make sense to mount the project dir inside the container instead of copying it... After the wheels are generated, the version string is fixed and that workaround is no longer needed...

Hm... what's really strange is that it works for me on my Mac. So `COPY` just deeply copies over the whole git repo. Maybe the git repo is already not correctly checkout in the Cirrus CI VM?

**Comment by FlorianWilhelm** [comment]
And it works 🎉  I don't know how many times I stumbled already over the [nested scaffolds problem](https://github.com/pyscaffold/pyscaffold/issues/544). 

What do you think? Can we merge it and should it be released as Version 0.7?

**Comment by abravalheri** [comment]
Wow! That was serious detective work!
Thank you very much @FlorianWilhelm.

I think it is a very welcome addition to the package in version 0.7!

**Pull request merged**